include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dyanmically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin);
+
+  // Context
+
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
+   * onNetworkNewConnection and onRequestHeaders() call onCreate().
+   * @param parent_context_id is the parent Context id for the context being created.  For a Network
+   * stream Context, this will be a Root Context id.
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  // Network
+
+  /**
+   * Call on a stream Context to indicate that a new network connection has been been created.
+   * Calls onStart().
+   */
+  virtual ProxyAction onNetworkNewConnection() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+  /**
+   * Call on a stream Context to indicate that data has arrived from downstream (e.g. on the
+   * incoming port that was accepted and for which the proxy is acting as a server).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onDownstreamData(int /* data_length */, bool /* end_of_stream */) {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+  /**
+   * Call on a stream Context to indicate that data has arrived from upstream (e.g. on the outgoing
+   * port that the proxy connected and for which the proxy is acting as a client).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onUpstreamData(int /* data_length */, bool /* end_of_stream */) {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /*
+   * The source of the close event.
+   * Unknown is when the source is not known.
+   * Local is when the close was initiated by the proxy.
+   * Remote is when the close was recieved from the peer.
+   */
+  enum class CloseType : uint32_t {
+    Unknown = 0,
+    Local = 1,
+    Remote = 2,
+  };
+
+  /**
+   * Call on a stream context to indicate that the downstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onDownstreamConnectionClose(CloseType /* close_type */) { unimplemented(); }
+
+  /**
+   * Call on a stream context to indicate that the upstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onUpstreamConnectionClose(CloseType /* close_type */) { unimplemented(); }
+
+  // HTTP
+
+  /**
+   * Call on a stream context to indicate that the request headers have arrived.  Calls onCreate().
```

Comment:
Can you clarify in these docs how the headers are obtained (same for trailers, metadata, etc.)

Reply:
Done.